ddtrace/llmobs/_integrations/langgraph.py

```diff
@@ -0,0 +1,174 @@
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.llmobs._constants import INPUT_VALUE
+from ddtrace.llmobs._constants import NAME
+from ddtrace.llmobs._constants import OUTPUT_VALUE
+from ddtrace.llmobs._constants import SPAN_KIND
+from ddtrace.llmobs._constants import SPAN_LINKS
+from ddtrace.llmobs._integrations.base import BaseLLMIntegration
+from ddtrace.llmobs._integrations.utils import format_langchain_io
+from ddtrace.llmobs._utils import _get_attr
+from ddtrace.llmobs._utils import _get_llmobs_parent_id
+from ddtrace.llmobs._utils import _get_nearest_llmobs_ancestor
+from ddtrace.span import Span
+
+
+class LangGraphIntegration(BaseLLMIntegration):
+    _integration_name = "langgraph"
+    _graph_nodes_by_task_id: Dict[str, Any] = {}  # maps task_id to dictionary of name, span, and span_links
+
+    def _llmobs_set_tags(
+        self,
+        span: Span,
+        args: List[Any],
+        kwargs: Dict[str, Any],
+        response: Optional[Any] = None,
+        operation: str = "",  # oneof graph, node
+    ):
+        if not self.llmobs_enabled:
+            return
+
+        inputs = get_argument_value(args, kwargs, 0, "input")
+        config = get_argument_value(args, kwargs, 1, "config", optional=True)
+        metadata = _get_attr(config, "metadata", {})
+        instance_id = metadata.get("langgraph_checkpoint_ns", "").split(":")[-1]
+        invoked_node = self._graph_nodes_by_task_id.setdefault(instance_id, {})
+        invoked_node["span"] = {"trace_id": "{:x}".format(span.trace_id), "span_id": str(span.span_id)}
+
+        span_links = [_default_span_link(span)]
+        invoked_node_span_links = invoked_node.get("span_links")
+        if invoked_node_span_links is not None:
+            span_links = invoked_node_span_links
+        current_span_links = span._get_ctx_item(SPAN_LINKS) or []
+
+        span._set_ctx_items(
+            {
+                SPAN_KIND: "agent" if operation == "graph" else "task",
+                INPUT_VALUE: format_langchain_io(inputs),
+                OUTPUT_VALUE: format_langchain_io(response),
+                NAME: self._graph_nodes_by_task_id.get(instance_id, {}).get("name") or kwargs.get("name", span.name),
+                SPAN_LINKS: current_span_links + span_links,
+            }
+        )
+        if operation == "graph" and not _is_subgraph(span):
+            self._graph_nodes_by_task_id.clear()
+
+    def llmobs_handle_pregel_loop_tick(
+        self, finished_tasks: dict, next_tasks: dict, more_tasks: bool, is_subgraph_node: bool = False
+    ):
+        """Compute incoming and outgoing span links between finished tasks and queued tasks in the graph."""
+        if not self.llmobs_enabled:
+            return
+        graph_span = (
+            tracer.current_span()
+        )  # we're running between nodes, so the current span should be the pregel graph
+        if graph_span is None or graph_span.span_type != SpanTypes.LLM:
+            return
```

Comment:
Just curious here - should this be an LLMOBS-specific span? If not, we would be creating a span link with a regular APM span right?
I think I could see a case if a user manually wraps/starts a span between the graph and node execution then we would break. Once the context management PR is merged, should we change this logic to specifically look for the active LLMObs span via `LLMObs._current_span()`?

Reply:
ah hmm good point. we do want it to be the current active llmobs span. i think i'll want to change this to `LLMObs._current_span()` once that lands. 

we'll still have the issue of manually generated spans though; all of this logic here is predicated on the only spans involved being langgraph-generated spans. but, i think we can utilize the beta-nature of this integration to see how exactly we'd wanna handle that case (i'm pretty sure manually generated spans would end up being disconnected linking-wise)

Reply:
Cool, since this is experimental for now I'm happy merging this as is and adding next steps once context management is merged + other misc fixes as we find them!